File: app/src/main/java/com/kotlinmockwebserver/example/app/HostSelectionInterceptor.java
package com.kotlinmockwebserver.example.app;

import java.io.IOException;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;

public final class HostSelectionInterceptor implements Interceptor {
    private volatile HttpUrl url;

    public void setHost(HttpUrl hostUrl) {
        this.url = hostUrl;
    }

    @Override public okhttp3.Response intercept(Chain chain) throws IOException {
        Request request = chain.request();
        HttpUrl url = this.url;
        if (url != null) {
            HttpUrl requestUrl = request.url().newBuilder()
                    .host(url.host())
                    .port(url.port())
                    .scheme(url.scheme())
                    .build();

            request = request.newBuilder()
                    .url(requestUrl)
                    .build();
        }
        return chain.proceed(request);
    }
}

File: app/src/androidTest/kotlin/com/kotlinmockwebserver/server_rule/interceptor/HostSelectionInterceptor.kt
package com.kotlinmockwebserver.server_rule.interceptor

import okhttp3.Interceptor
import java.io.IOException

class HostSelectionInterceptor : Interceptor {
    @Volatile private var host: String? = null

    fun setHost(host: String) {
        this.host = host
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): okhttp3.Response {
        var request = chain.request()
        val host = this.host
        if (host != null) {
            val newUrl = request.url().newBuilder()
                    .host(host)
                    .build()
            request = request.newBuilder()
                    .url(newUrl)
                    .build()
        }
        return chain.proceed(request)
    }
}
